src/internal/m365/collection/drive/collections.go

```diff
@@ -688,232 +670,297 @@ func (c *Collections) getCollectionPath(
 		}
 
-		var (
-			itemID   = ptr.Val(item.GetId())
-			itemName = ptr.Val(item.GetName())
-			isFolder = item.GetFolder() != nil || item.GetPackageEscaped() != nil
-			ictx     = clues.Add(
-				ctx,
-				"item_id", itemID,
-				"item_name", clues.Hide(itemName),
-				"item_is_folder", isFolder)
-		)
-
-		if item.GetMalware() != nil {
-			addtl := graph.ItemInfo(item)
-			skip := fault.FileSkip(fault.SkipMalware, driveID, itemID, itemName, addtl)
+		if reset {
+			newPrevPaths = map[string]string{}
+			currPrevPaths = map[string]string{}
+			c.CollectionMap[driveID] = map[string]*Collection{}
+			invalidPrevDelta = true
+		}
 
-			if isFolder {
-				skip = fault.ContainerSkip(fault.SkipMalware, driveID, itemID, itemName, addtl)
+		for _, item := range page {
+			if el.Failure() != nil {
+				break
 			}
 
-			errs.AddSkip(ctx, skip)
-			logger.Ctx(ctx).Infow("malware detected", "item_details", addtl)
-
-			continue
-		}
-
-		// Deleted file or folder.
-		if item.GetDeleted() != nil {
-			if err := c.handleDelete(
-				itemID,
+			err := c.processItem(
+				ctx,
+				item,
 				driveID,
+				driveName,
 				oldPrevPaths,
 				currPrevPaths,
 				newPrevPaths,
-				isFolder,
-				excluded,
-				invalidPrevDelta); err != nil {
-				return nil, clues.Stack(err).WithClues(ictx)
+				excludedItemIDs,
+				invalidPrevDelta,
+				el)
+			if err != nil {
+				el.AddRecoverable(ctx, clues.Stack(err))
 			}
-
-			continue
 		}
+	}
 
-		collectionPath, err := c.getCollectionPath(driveID, item)
-		if err != nil {
-			el.AddRecoverable(ctx, clues.Stack(err).
-				WithClues(ictx).
-				Label(fault.LabelForceNoBackupCreation))
+	du, err := pager.Results()
+	if err != nil {
+		return du, nil, clues.Stack(err)
+	}
 
-			continue
-		}
+	return du, newPrevPaths, el.Failure()
+}
 
-		// Skip items that don't match the folder selectors we were given.
-		if shouldSkip(ctx, collectionPath, c.handler, driveName) {
-			logger.Ctx(ictx).Debugw("path not selected", "skipped_path", collectionPath.String())
-			continue
+func (c *Collections) processItem(
+	ctx context.Context,
+	item models.DriveItemable,
+	driveID, driveName string,
+	oldPrevPaths, currPrevPaths, newPrevPaths map[string]string,
+	excluded map[string]struct{},
+	invalidPrevDelta bool,
+	skipper fault.AddSkipper,
+) error {
+	var (
+		itemID   = ptr.Val(item.GetId())
+		itemName = ptr.Val(item.GetName())
+		isFolder = item.GetFolder() != nil || item.GetPackageEscaped() != nil
+		ictx     = clues.Add(
+			ctx,
+			"item_id", itemID,
+			"item_name", clues.Hide(itemName),
+			"item_is_folder", isFolder)
+	)
+
+	if item.GetMalware() != nil {
+		addtl := graph.ItemInfo(item)
+		skip := fault.FileSkip(fault.SkipMalware, driveID, itemID, itemName, addtl)
+
+		if isFolder {
+			skip = fault.ContainerSkip(fault.SkipMalware, driveID, itemID, itemName, addtl)
 		}
 
-		switch {
-		case isFolder:
-			// Deletions are handled above so this is just moves/renames.
-			var prevPath path.Path
+		skipper.AddSkip(ctx, skip)
```

Comment:
honestly this feels like a bit of an anti-pattern. We're taking in (a part of) the fault bus, but only using it to add skips. Errors are returned like normal error handling with the assumption that the caller will add them to the bus

It feels like either this function should handle all the necessary steps to add stuff to the bus, both errors and skips, or the caller should

Reply:
Grokked and aggreed.  I think I'm going to hold off on changes further here within this PR, since I'd like to revisit this code with some additional refactoring and this PR is already doing enough.  It'll be easier for me to work on design improvements there.  But I will get to it.